src/main/cpp/lang/xml.cpp

```diff
@@ -0,0 +1,478 @@
+#include "ulight/impl/lang/xml.hpp"
+#include "ulight/impl/lang/html.hpp"
+#include "ulight/impl/lang/xml_chars.hpp"
+
+#include "ulight/impl/buffer.hpp"
+#include "ulight/impl/highlight.hpp"
+#include "ulight/impl/highlighter.hpp"
+
+#include "ulight/impl/unicode_algorithm.hpp"
+#include "ulight/ulight.hpp"
+
+#include <cctype>
+#include <cmath>
+#include <functional>
+#include <iterator>
+#include <string>
+
+namespace ulight {
+
+namespace xml {
+
+namespace {
+
+constexpr std::u8string_view comment_prefix = u8"<!--";
+constexpr std::u8string_view comment_suffix = u8"-->";
+constexpr std::u8string_view illegal_comment_sequence = u8"--";
+constexpr std::u8string_view cdata_section_prefix = u8"<![CDATA[";
+constexpr std::u8string_view cdata_section_suffix = u8"]]>";
+
+[[nodiscard]]
+bool contains_xml_string(std::u8string_view str)
+{
+
+    if (str.length() < 3) {
+        return false;
+    }
+
+    for (std::size_t i = 0; i <= str.length() - 3; i++) {
+        if ((str[i] == u8'x' || str[i] == u8'X') && //
+            (str[i + 1] == u8'm' || str[i + 1] == u8'M') && //
+            (str[i + 2] == u8'l' || str[i + 2] == u8'L')) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+} // namespace
+
+[[nodiscard]]
+std::size_t match_whitespace(std::u8string_view str)
+{
+    const std::size_t result = utf8::find_if_not(str, is_xml_whitespace_character);
+
+    return result == std::string::npos ? str.length() : result;
+}
+
+[[nodiscard]]
+Name_Match_Result match_name_permissive(
+    std::u8string_view str,
+    const std::function<bool(std::u8string_view)>& is_stop_sequence
+)
+{
+    if (str.empty()) {
+        return {};
+    }
+
+    Name_Match_Result result;
+
+    if (!is_xml_name_start_character(str[0])) {
+        result.error_indicies.insert(0);
+    }
+
+    result.length++;
+    while (result.length < str.length() && !is_stop_sequence(str.substr(result.length))) {
+        if (!is_xml_name_character(str[result.length])) {
+            result.error_indicies.insert(result.length);
+        }
+
+        result.length++;
+    }
+
+    return result;
+}
+
+[[nodiscard]]
+std::size_t match_text(std::u8string_view str)
+{
+    const std::size_t result
+        = utf8::find_if(str, [](char32_t c) { return c == u8'<' || c == u8'&'; });
+
+    return result == std::string::npos ? str.length() : result;
```

Comment:
```suggestion
return ascii::length_if_not(str, [](char8_t c) { return c == u8'<' || c == u8'&'; });
```
When an algorithm doesn't require full UTF-8 decoding (and looking for an ASCII character doesn't), the `ascii::` algorithms should be used instead.

Another way to implement it would be:
```suggestion
const std::size_t result = str.find_first_of(u8"<&");
return result == std::u8string_view::npos ? str.length() : result;
```
Also note the use of `std::u8string_view::npos` instead of `std::string::npos`. While they are the same in practice, it doesn't make much sense to rely on that.